### PR TITLE
Offline Pages : Scroll To Page Once Edited

### LINKED ISSUE #11132: Scroll to a page after leaving the editor
When the user eg. publishes a page draft, they are redirected back to the "Draft" tab. However, the page they just edited disappears (it's moved to the "Published" tab). If they manually go to "Published" tab, the page might not be visible at the top as pages are ordered alphabetically. 
Goal of this task is to automatically change the tab and scroll to the page after the user leaves the editor.

### PR DESCRIPTION
**Pull request by jd-alexander**
Fixes #11132

## Findings
There was already behavior for scrolling so that was leveraged to make scrolling possible when pages are modified. 

## Solution
1. The `scrollToPage` LiveData object was triggered when page edits are finished and when a page's status is changed. This was done since those were the times when a user would be interacting with the `Pages List` and not know the location of a post they performed an action on. 

2. I switched the use of `remoteId` for `localId` for the page query since pages that haven't been uploaded don't have a `remoteId` and scroll wasn't working for them. 

3. I realized that scrolling to an item before I am sure the list data was loaded in the `ViewModel` was sometimes causing scrolls to not work, so I set the scroll to retry since we know the page exists and there could always be timing issues. This seemed like the most sensible solution without disrupting the existing scroll behavior. 

4. Something I have realized with this solution is it's easier to identify a page when it's being uploaded since the progress bar will be visible but with a page status change, you have to find the page in the screen port since it's position won't be obvious. This might only be happening because I am testing, as others will have more meaningful page names and such forth, so this can be overlooked. I experimented with `scaling` and modifying the `alpha` using animations to add a subtle hint for the page but it didn't work well since the page is being invalidated a lot and to prevent this behavior, a modification would have to be made to the adapter which isn't necessary at this point in time. 

5. Sometimes the scroll didnt't complete because during the smooth scroll animation the list is invalidated. I think this happens because within the `changePageStatus` function, the page is refreshed and the scroll is initiated and then when the action is completed a `reloadPages` function was called. I removed this f328d58 since the database was updated and I didn't see why the pages needed to be reloaded. This removal is somewhat similar to the removal from the `postUploadStarted` function you did when cleaning up the `PageEventListener`.  Let me know what you think about this. 

6. There are a few times, I have noticed that after saving a draft, it doesn't scroll to the location. I am still trying to figure out why that happens. 

## Testing

1. Go to my site. 
2. Go to Pages. 
3. Change the page status of pages on the Published, Drafts,  Scheduled, Trashed Tab. 
4. The app should then jump to that tab and the page should be visible. 
---------------
1. Go to my site. 
2. Go to Pages. 
3. Create a new page or update a page, especially by scheduling it.
4. Publish or save the page.
5. The app should then jump to that tab and the page should be visible. 

## Reviewing 
Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [x] I have considered adding accessibility improvements for my changes.
- [x] If it's feasible, I have added unit tests.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt [hunk 1]
```diff
@@ -147,12 +148,17 @@ class PageListViewModel @Inject constructor(
...
-        val position = _pages.value?.indexOfFirst { it is Page && it.remoteId == remotePageId } ?: -1
+    fun setPageToScrollTo(pageId: LocalId) {
+        pageToScrollTo = pageId
+    }
+
+    private fun onScrollToPageRequested(localPageId: Int) {
+        val position = _pages.value?.indexOfFirst { it is Page && it.localId == localPageId } ?: -1
         if (position != -1) {
             _scrollToPosition.postValue(position)
+            pageToScrollTo = null
```
Comment (jd-alexander) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: I am clearing the `pageToScrollTo` here only when the id was found due to times when the scroll wouldn't take place. I am not sure about this and it's mutating state in a different location so I need your thoughts on this and I will double-check it. Didn't want it to stop the review though.

**Comment by oguzkocer** [comment]
We're freezing `14.3` today, so this PR is being bumped to `14.4`. If you'd like it to ship with `14.3`, please merge it into `release/14.3` and ping me – I'll be happy to cut a new beta release.

**Comment by malinajirka** [comment]
Thanks @jd-alexander! The code looks good overall. I've tested the app and I haven't found any new issues.

>Sometimes the scroll didnt't complete because during the smooth scroll animation the list is invalidated. I think this happens because within the changePageStatus function, the page is refreshed and the scroll is initiated and then when the action is completed a reloadPages function was called. I removed this f328d58 since the database was updated and I didn't see why the pages needed to be reloaded. This removal is somewhat similar to the removal from the postUploadStarted function you did when cleaning up the PageEventListener. Let me know what you think about this.


I think the removal is ok. How would you feel about using `scrollTo` instead of `smoothScrollTo`? We might want to ask a designer as it's purely a design decision. 

>Something I have realized with this solution is it's easier to identify a page when it's being uploaded since the progress bar will be visible but with a page status change, you have to find the page in the screen port since it's position won't be obvious. This might only be happening because I am testing, as others will have more meaningful page names and such forth, so this can be overlooked. I experimented with scaling and modifying the alpha using animations to add a subtle hint for the page but it didn't work well since the page is being invalidated a lot and to prevent this behavior, a modification would have to be made to the adapter which isn't necessary at this point in time.


Yeah, I agree this isn't optimal. The root cause is that the PageList doesn't use the UploadService for uploading the status change -> therefore the uploading progress state doesn't get propagated. 

We might want to consider using either
`LinearLayoutManager.scrollToPositionWithOffset(position, 0)`
or a custom SmoothScroller

```
               val scroller = object : LinearSmoothScroller(context) {
                    override fun getVerticalSnapPreference(): Int {
                        return LinearSmoothScroller.SNAP_TO_START
                    }
                }
                ...
                smoothScroller.targetPosition = position
                recyclerView.layoutManager?.startSmoothScroll(smoothScroller)
```
So the item is always at the top? Wdyt?

**Comment by jd-alexander** [comment]
Wow @malinajirka the custom SmoothScroller works like magic 😄 Thanks for this. This resolves all the concerns I had about the scrolling and knowing where to look since it will always be at the top!

**Comment by oguzkocer** [comment]
We're freezing `14.4` today, so this PR is being bumped to `14.5`. If you'd like it to ship with 14.4, please merge it into release/14.4 and ping me – I'll be happy to cut a new beta release.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt [hunk 2]
```diff
@@ -475,6 +476,14 @@ class PagesViewModel
 
     private fun publishPageNow(remoteId: Long) {
         _publishAction.value = pageMap[remoteId]
+        launch(uiDispatcher) {
+            withContext(defaultDispatcher) {
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: I think we can remove `withContext(defaultDispatcher)` as delay takes care of not blocking the current thread. Wdyt?

Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Sure! Sounds like a good idea!

**Comment by jd-alexander** [comment]
Hi @malinajirka this is ready for another review!
> - When I click on "Publish Now" on a draft page list item, it doesn't change tabs and scroll to the item

This was resolved in b86b3a1 02a9541

>- When I open page list and I move an item to drafts it changes the tab but it doesn't scroll 

This was resolved in b5457ba

**Comment by malinajirka** [comment]
>When I open page list and I move an item to drafts it changes the tab but it doesn't scroll
>>This was resolved in b5457ba

I'm not sure about this solution. It works, but I'm a bit worried it'll have performance impact on low end devices. We have static number of tabs/fragments, but the fragments themselves are not static. It seems the scrolling issue isn't vital and reducing performance might not be worth it. We might want to consider reverting the change in this PR and discuss it somewhere else so we don't block this PR. Wdyt? (Would for example ordering pages in "Trashed" tab by their modified date solve the issue?)

>Also I have a question. When a page status is changed we are updating the pageMap that's in memory i.e the list will already has the most up to date page with it's updated status so is it redundant to update the pageMap from the database again?

Hmm agree, seems like it's redundant.

>When I click on "Publish Now" on a draft page list item, it doesn't change tabs and scroll to the item
>>This was resolved in b86b3a1 02a9541

Awesome, thank you!

**Comment by jd-alexander** [comment]
> I'm not sure about this solution. It works, but I'm a bit worried it'll have performance impact on low end devices. We have static number of tabs/fragments, but the fragments themselves are not static. It seems the scrolling issue isn't vital and reducing performance might not be worth it. We might want to consider reverting the change in this PR and discuss it somewhere else so we don't block this PR. Wdyt? (Would for example ordering pages in "Trashed" tab by their modified date solve the issue?)

Yes, this seems like a good idea since finding the optimal solution is proving itself a bit difficult at the time.

**Pull request merged**